crates/sui-tool/src/db_tool/mod.rs

```diff
@@ -135,6 +189,92 @@ pub fn print_db_duplicates_summary(db_path: PathBuf) -> anyhow::Result<()> {
     Ok(())
 }
 
+pub fn print_transaction_checkpoint(
+    path: &Path,
+    opt: PrintTransactionOptions,
+) -> anyhow::Result<()> {
+    let perpetual_db = AuthorityPerpetualTables::open(&path.join("store"), None);
+    let Some((epoch, checkpoint_seq_num)) = perpetual_db.get_checkpoint_sequence_number(&opt.digest)? else {
+        bail!("Transaction {:?} not executed in a checkpoint!", opt.digest);
+    };
+    println!(
+        "Transaction {:?} executed in epoch {} checkpoint {}",
+        opt.digest, epoch, checkpoint_seq_num
+    );
+    Ok(())
+}
+
+// Force removes a transaction and its outputs, if no other dependent transaction has executed yet.
+// Usually this should be paired with rewind_checkpoint_execution() to re-execute the removed
+// transaction, to repair corrupted database.
+// Dry run with: cargo run --package sui-tool -- db-tool --db-path /opt/sui/db/authorities_db/live remove-transaction --digest xxxx
+// Add --confirm to actually remove the transaction.
+pub fn remove_transaction(path: &Path, opt: RemoveTransactionOptions) -> anyhow::Result<()> {
+    let perpetual_db = AuthorityPerpetualTables::open(&path.join("store"), None);
+    let Some(_transaction) = perpetual_db.get_transaction(&opt.digest)? else {
+        bail!("Transaction {:?} not found and cannot be re-executed!", opt.digest);
+    };
+    let Some(effects) = perpetual_db.get_effects(&opt.digest)? else {
+        bail!("Transaction {:?} not executed or effects have been pruned!", opt.digest);
+    };
+    let mut objects_to_remove = vec![];
+    for mutated_obj in effects.modified_at_versions() {
+        let new_objs = perpetual_db.get_newer_object_keys(mutated_obj)?;
+        if new_objs.len() > 1 {
+            bail!(
+                "Dependents of transaction {:?} have already executed! Mutated object: {:?}, new objects: {:?}",
+                opt.digest,
+                mutated_obj,
+                new_objs,
+            );
```

Comment:
We can but I don't think it is necessary. Using this logic is very subtle already and whoever runs it needs to fully understand the risk. There can be many more ways for this logic to corrupt db, and we will not be able to prevent them all.